conjureplugin/config/internal/v1/config.go

```diff
@@ -82,10 +88,118 @@ func (cfg *IRLocatorConfig) UnmarshalYAML(unmarshal func(interface{}) error) err
+		SkipDeleteGeneratedFiles: true,
+	}
+
+	normalizedOutput := filepath.Clean(v1proj.OutputDir)
+	v2Default := filepath.Clean(filepath.Join(v2.DefaultOutputDir, projectName))
+
+	if normalizedOutput == v2Default {
+		// v1 output matches v2 default (internal/generated/conjure/{projectName}).
+		// Leave OutputDir empty to use v2's default, with OmitTopLevelProjectDir=false.
+	} else if normalizedOutput == filepath.Clean(v2.DefaultOutputDir) {
+		// v1 output is internal/generated/conjure (no project subdirectory).
+		// Leave OutputDir empty to use v2's default, but set OmitTopLevelProjectDir=true.
+		v2proj.OmitTopLevelProjectDir = true
+	} else if normalizedOutput == filepath.Clean(projectName) {
+		// v1 output is just the project name as a directory.
+		// Set OutputDir to "." and don't omit project dir to get ./{projectName}.
+		v2proj.OutputDir = "."
+	} else {
+		// v1 output is a custom path that doesn't match any v2 convention.
+		// Set OutputDir to the custom path and omit the project subdirectory.
+		v2proj.OutputDir = normalizedOutput
+		v2proj.OmitTopLevelProjectDir = true
+	}
+
+	return v2proj
+}
+
+// ToV2 intelligently converts a v1 config to v2, attempting to use v2 defaults when possible.
+// This is the conversion logic used by UpgradeConfig.
+func (v1cfg *ConjurePluginConfig) ToV2() v2.ConjurePluginConfig {
+	// Create v2 config with intelligent field mapping
+	v2cfg := v2.ConjurePluginConfig{
+		ConfigWithVersion: versionedconfig.ConfigWithVersion{Version: "2"},
+		GroupID:           v1cfg.GroupID,
+		ProjectConfigs:    make(map[string]v2.SingleConjureConfig),
+	}
+
+	// Convert each project using the per-project conversion logic
+	for projectName, v1proj := range v1cfg.ProjectConfigs {
+		v2cfg.ProjectConfigs[projectName] = v1proj.ToV2(projectName)
+	}
+
+	// Check if we need to allow conflicting output directories.
+	// We detect conflicts the same way ToParams does: exact same directory AND parent-child relationships.
+	// Calculate the actual output directories for each project
+	outputDirs := make(map[string][]string)
+	for projectName, proj := range v2cfg.ProjectConfigs {
+		resolvedOutputDir := proj.ResolvedOutputDir(projectName)
+		outputDirs[resolvedOutputDir] = append(outputDirs[resolvedOutputDir], projectName)
+	}
+
+	v2cfg.AllowConflictingOutputDirs = len(GetConflictingOutputDirs(outputDirs)) > 0
+
+	return v2cfg
+}
+
 func UpgradeConfig(cfgBytes []byte) ([]byte, error) {
 	var cfg ConjurePluginConfig
 	if err := yaml.UnmarshalStrict(cfgBytes, &cfg); err != nil {
 		return nil, errors.Wrapf(err, "failed to unmarshal conjure-plugin v1 configuration")
 	}
+
+	cfgBytes, err := yaml.Marshal(cfg.ToV2())
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to marshal upgraded v2 configuration")
+	}
+
 	return cfgBytes, nil
 }
+
+func GetConflictingOutputDirs(outputDirToProjects map[string][]string) []error {
+	var warnings []error
+
+	sortedOutputDir := slices.Sorted(maps.Keys(outputDirToProjects))
```

Comment:
Pluralize variable name to make clear it is a slice/collection

```suggestion
	sortedOutputDirs := slices.Sorted(maps.Keys(outputDirToProjects))
```